### PR TITLE
Adding ATF support to P2P references

### PR DESCRIPTION
**Pull request by mishra14**
## Bug
Currently ATF does not work in P2P references. 

## Fix
Details: For checking P2P compatibility, we match referenced project framework with the project framework using `FrameworkReducer`. But the `FrameworkReducer` currently does not handle `AssetTargetFramework` as commented [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs#L17) because it is used for other compat checks. So this PR makes [PackageSpecReferenceDependencyProvider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs#L188) more robust by checking if no compatible framework was found and if another check is needed with `AssetTargetFramework`

## Testing/Validation
Tests Added: Yes
Validation done:  Manual validation

### CODE REVIEW

File: src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs [hunk 1]
```diff
@@ -69,6 +69,24 @@ public NuGetFramework GetNearest(NuGetFramework framework, IEnumerable<NuGetFram
                 }
             }
 
+
+            if (nearest == null || nearest.IsAny)
+            {
+                // Try ATF
+                var atfFallbackFramework = framework as AssetTargetFallbackFramework;
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs]:
```
/// <summary>
    /// AssetTargetFallbackFramework only fallback when zero assets are selected. These do not 
    /// auto fallback during GetNearest as FallbackFramework would.
 /// </summary>
```

This is the description mentioned for `AssetTargetFallbackFramework` do we know why this is explicitly mentioned not to fallback during GetNearest?

**Review by jainaashish — CHANGES_REQUESTED** [review]
As commented, looks like we intentionally skipped falling to `ATF` during `GetNearest` api, please get confirmation before proceeding with this PR.

File: src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs [hunk 2]
```diff
@@ -52,11 +52,11 @@ public NuGetFramework GetNearest(NuGetFramework framework, IEnumerable<NuGetFram
         {
             var nearest = GetNearestInternal(framework, possibleFrameworks);
 
-            var fallbackFramework = framework as FallbackFramework;
-
-            if (fallbackFramework != null)
+            if (nearest == null || nearest.IsAny)
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs]: Combine the PTF/ATF checks and loops

1. Check if framework is ATF, read fallbacks
1. If not ATF check if PTF, read fallbacks
1. If fallbacks exist loop through them and check compat (doesn't matter where they came from)

**Comment by mishra14** [comment]
ping @jainaashish @emgarten 

I hav eupdated the PR to make [PackageSpecReferenceDependencyProvider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs#L188) more robust by checking if no compatible framework was found and if another check is needed with `AssetTargetFramework`

**Review by jainaashish — APPROVED** [review]
LGTM

File: src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Frameworks/FrameworkReducer.cs]: revert this.

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/FrameworkReducerTests.cs [hunk 4]
```diff
@@ -1068,4 +1068,4 @@ public void FrameworkReducer_GetNearestAspNeg(string project, string framework)
             Assert.Null(result);
         }
     }
-}
+}
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Frameworks.Test/FrameworkReducerTests.cs]: revert this.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 5]
```diff
@@ -4810,6 +4812,97 @@ public void RestoreNetCore_NETCore_ProjectToProject_IgnoreMissingProjectReferenc
...
+                // Act
+                var r = Util.RestoreSolution(pathContext);
+
+                // Assert
+                r.Success.Should().BeTrue();
+                foreach (var graph in projectA.AssetsFile.Targets)
+                {
+                    var lib = graph.GetTargetLibrary("b");
+
+                    lib.CompileTimeAssemblies.Select(e => e.Path)
```
Comment (nkolev92) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:4853]: I'm confused...how can this start with bin?

**Review by nkolev92 — APPROVED** [review]
LGTM.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs (hunk 5 above)
Comment (mishra14) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs:4853]: I think this comes from [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs#L486). The test projects are more like facades and dont really have files, which may be causing this. But thats a general test issue.

**Review by emgarten — APPROVED** [review]
🚀

**Pull request merged**